lib/order.rb

```diff
@@ -7,12 +11,51 @@ def initialize(id, products)
       @products = products
     end
 
+    # sums product prices, adds 7.5% tax, and rounds to two decimal places
     def total
-      # TODO: implement total
+      subtotal = 0
+      @products.each do |name, price|
+        subtotal += price
+      end
+      total = (subtotal * 1.075).round(2)
+      return total
     end
 
+    # returns false if the product name is already included in the order's products,
+    # adds product to the instance's product hash if not already present and returns true
     def add_product(product_name, product_price)
-      # TODO: implement add_product
+      if @products.keys.include?(product_name)
+        return false
+      else
+        @products[product_name] = product_price
+        return true
+      end
+    end
+
+    # reads csv file (set to orders.csv by default if no other file is given)
+    # iterates through each line (array) of the csv file to parse the id and product data
+    # for each line, creates a new instance of the Order class and adds it to an array of all Order instances
+    def self.all(csv_file=ORDER_FILE)
+      csv_array = CSV.read(csv_file, 'r')
+      all_orders = []
+      csv_array.each do |order|
+        id = order[0].to_i
+        products = Hash[order[1].split(/:|;/).each_slice(2).collect { |k, v| [k,v.to_f] }]
+        new_order = Order.new(id, products)
```

Comment:
Good work getting this tricky logic sorted out. Line 43 definitely works, but it's not easy to read. Often it's helpful to break a complex piece of work like this into multiple steps, storing intermediate results in local variables.